tick/eval/node_value_acessors.go

```diff
@@ -0,0 +1,547 @@
+	return "", ErrTypeGuardFailed
+}
+
+func (n *EvalNumberNode) BoolValue(scope *tick.Scope, executionAux ExecutionAux) (bool, error) {
+	return false, ErrTypeGuardFailed
+}
+func (n *EvalNumberNode) RegexValue(scope *tick.Scope, executionAux ExecutionAux) (*regexp.Regexp, error) {
+	return nil, ErrTypeGuardFailed
+}
+
+//-----------------------------------------
+// RegexNode NodeValueAcessor Implementation
+
+func (n *EvalRegexNode) Value(scope *tick.Scope, executionAux ExecutionAux) (interface{}, error) {
+	return n.Node.Regex, nil
+}
+
+func (n *EvalRegexNode) RegexValue(scope *tick.Scope, executionAux ExecutionAux) (*regexp.Regexp, error) {
+	return n.Node.Regex, nil
+}
+
+func (n *EvalRegexNode) StringValue(scope *tick.Scope, executionAux ExecutionAux) (string, error) {
+	return "", ErrTypeGuardFailed
+}
+
+func (n *EvalRegexNode) Float64Value(scope *tick.Scope, executionAux ExecutionAux) (float64, error) {
+	return float64(0), ErrTypeGuardFailed
+}
+
+func (n *EvalRegexNode) Int64Value(scope *tick.Scope, executionAux ExecutionAux) (int64, error) {
+	return int64(0), ErrTypeGuardFailed
+}
+
+func (n *EvalRegexNode) BoolValue(scope *tick.Scope, executionAux ExecutionAux) (bool, error) {
+	return false, ErrTypeGuardFailed
+}
+
+//-----------------------------------------
+// StringNode NodeValueAcessor Implementation
+
+func (n *EvalStringNode) Value(scope *tick.Scope, executionAux ExecutionAux) (interface{}, error) {
+	return n.Node.Literal, nil
+}
+
+func (n *EvalStringNode) StringValue(scope *tick.Scope, executionAux ExecutionAux) (string, error) {
+	return n.Node.Literal, nil
+}
+
+func (n *EvalStringNode) Float64Value(scope *tick.Scope, executionAux ExecutionAux) (float64, error) {
+	return float64(0), ErrTypeGuardFailed
+}
+
+func (n *EvalStringNode) Int64Value(scope *tick.Scope, executionAux ExecutionAux) (int64, error) {
+	return int64(0), ErrTypeGuardFailed
+}
+
+func (n *EvalStringNode) BoolValue(scope *tick.Scope, executionAux ExecutionAux) (bool, error) {
+	return false, ErrTypeGuardFailed
+}
+
+func (n *EvalStringNode) RegexValue(scope *tick.Scope, executionAux ExecutionAux) (*regexp.Regexp, error) {
+	return nil, ErrTypeGuardFailed
+}
+
+//-----------------------------------------
+// BinaryNode NodeValueAcessor Implementation
+//
+// it is safe to assume in BinaryNode speicific value functions to use EvalNum/EvalBool
+// because we are using getConstantNodeType to do type inference, and it's really accurate
+
+func (n *EvalBinaryNode) Value(scope *tick.Scope, executionAux ExecutionAux) (interface{}, error) {
+	switch getConstantNodeType(n.Node) {
+	case TInt64:
+		return n.Int64Value(scope, executionAux)
+	case TFloat64:
+		return n.Float64Value(scope, executionAux)
+	case TBool:
+		return n.BoolValue(scope, executionAux)
+	default:
+		return nil, nil
+	}
+}
+
+func (n *EvalBinaryNode) RegexValue(scope *tick.Scope, executionAux ExecutionAux) (*regexp.Regexp, error) {
+	return nil, nil
```

Comment:
ErrTypeGuardFailed doesn't make sense in this case, there is no case when the BinaryNode won't return bool/float/int unless it contains invalid operator.

I will change the nil of the error to be an explanative error